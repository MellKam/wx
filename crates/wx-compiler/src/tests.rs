use super::*;
use indoc::indoc;
use insta::{assert_binary_snapshot, assert_yaml_snapshot};
use string_interner::{StringInterner, backend::StringBackend};

#[allow(unused)]
struct TestCase {
    files: files::Files,
    interner: StringInterner<StringBackend>,
    ast: ast::parser::ParserResult,
    hir: hir::builder::BuilderResult,
    mir: mir::MIR,
    module: wasm::Module,
    bytecode: Vec<u8>,
}

impl TestCase {
    fn assert_snapshot(name: &str, source: &str) {
        let mut interner = StringInterner::new();
        let mut files = files::Files::new();
        let file_id = files
            .add(name.to_string() + ".wx", source.to_string())
            .unwrap();

        let ast =
            ast::parser::Parser::parse(file_id, &files.get(file_id).unwrap().source, &mut interner);
        assert_yaml_snapshot!(name.to_string() + "_ast", ast);

        if ast.diagnostics.len() > 0 {
            return;
        }

        let hir = hir::Builder::build(&ast.ast, &mut interner);
        assert_yaml_snapshot!(name.to_string() + "_hir", hir);

        let mir = mir::Builder::build(&hir.hir);
        assert_yaml_snapshot!(name.to_string() + "_mir", mir);

        let module = wasm::Builder::build(&mir).unwrap();
        assert_yaml_snapshot!(name.to_string() + "_wasm", module);

        let bytecode = wasm::Encoder::encode(&module, &interner);
        assert_binary_snapshot!((name.to_string() + ".wasm").as_ref(), bytecode.clone());
    }
}

#[test]
fn factorial() {
    TestCase::assert_snapshot(
        "factorial",
        indoc! {"
        export func factorial(n: i32): i32 {
            if n <= 1 { return 1 };
            n * factorial(n - 1)
        }
    "},
    );
}

#[test]
fn pow() {
    TestCase::assert_snapshot(
        "pow",
        indoc! {"
        export func pow(base: i32, mut exp: i32): i32 {
            if exp < 0 {
                return 0; 
            };
            
            local mut result: i32 = 1;
            loop {
                if exp == 0 {
                break result; 
                };
                
                result *= base;
                exp -= 1;
            }
        }
    "},
    );
}

#[test]
fn func_pointers() {
    TestCase::assert_snapshot(
        "func_pointers",
        indoc! {"
        func add(a: i32, b: i32): i32 { a + b }
        func sub(a: i32, b: i32): i32 { a - b }

        func apply(
            binop: func(i32, i32) -> i32, 
            a: i32, 
            b: i32,
        ): i32 {
            binop(a, b)
        }

        export func main(): i32 {
            local a = apply(add, 5, 10);
            local b = apply(sub, 10, 5);
            local c: i32 = 5;

            a + b
        }
    "},
    );
}
