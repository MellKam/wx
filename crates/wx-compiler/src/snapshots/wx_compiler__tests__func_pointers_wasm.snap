---
source: crates/wx-compiler/src/tests.rs
expression: module
---
types:
  signatures:
    - param_count: 2
      param_results:
        - I32
        - I32
        - I32
    - param_count: 3
      param_results:
        - I32
        - I32
        - I32
        - I32
    - param_count: 0
      param_results:
        - I32
globals:
  globals: []
tables:
  tables:
    - ty: FuncRef
      limits:
        Initial: 2
elements:
  segments:
    - table_index: 0
      offset: 0
      indices:
        - 0
        - 1
functions:
  types:
    - 0
    - 0
    - 1
    - 2
exports:
  items:
    - Function:
        name: 8
        func_index: 3
code:
  expressions: []
  functions:
    - name: 1
      locals:
        - name: 2
          ty: I32
        - name: 4
          ty: I32
      expressions:
        - I32Add:
            left:
              LocalGet:
                local_index: 0
            right:
              LocalGet:
                local_index: 1
    - name: 5
      locals:
        - name: 2
          ty: I32
        - name: 4
          ty: I32
      expressions:
        - I32Sub:
            left:
              LocalGet:
                local_index: 0
            right:
              LocalGet:
                local_index: 1
    - name: 6
      locals:
        - name: 7
          ty: I32
        - name: 2
          ty: I32
        - name: 4
          ty: I32
      expressions:
        - CallIndirect:
            expr:
              LocalGet:
                local_index: 0
            table_index: 0
            type_index: 0
            arguments:
              - LocalGet:
                  local_index: 1
              - LocalGet:
                  local_index: 2
    - name: 8
      locals:
        - name: 2
          ty: I32
        - name: 4
          ty: I32
        - name: 9
          ty: I32
      expressions:
        - LocalSet:
            local_index: 0
            value:
              Call:
                function: 2
                arguments:
                  - I32Const:
                      value: 0
                  - I32Const:
                      value: 5
                  - I32Const:
                      value: 10
        - LocalSet:
            local_index: 1
            value:
              Call:
                function: 2
                arguments:
                  - I32Const:
                      value: 1
                  - I32Const:
                      value: 10
                  - I32Const:
                      value: 5
        - LocalSet:
            local_index: 2
            value:
              I32Const:
                value: 5
        - I32Add:
            left:
              LocalGet:
                local_index: 0
            right:
              LocalGet:
                local_index: 1
